src/elements/element.line.js

```diff
@@ -52,56 +33,91 @@ function setStyle(ctx, vm) {
 }
 
-function normalPath(ctx, points, spanGaps, vm) {
-	const steppedLine = vm.steppedLine;
-	const lineMethod = steppedLine ? helpers.canvas._steppedLineTo : helpers.canvas._bezierCurveTo;
-	let move = true;
-	let index, currentVM, previousVM;
+function lineTo(ctx, previous, target) {
+	ctx.lineTo(target.x, target.y);
+}
+
+function getLineMethod(line) {
+	return line.steppedLine ? _steppedLineTo
+		: !line.tension ? lineTo
+		: _bezierCurveTo;
+}
+
+function getInterpolationMethod(line) {
+	return line.steppedLine ? steppedInterpolation
+		: !line.tension ? pointInLine
+		: bezierInterpolation;
+}
 
-	for (index = 0; index < points.length; ++index) {
-		currentVM = points[index]._view;
+function pathSegment(ctx, line, segment, options) {
+	const {start, end, loop} = segment;
+	const {points, _view: vm} = line;
+	const lineMethod = getLineMethod(vm);
+	const count = points.length;
+	let {move = true, reverse} = options || {};
+	let ilen = end < start ? count + end - start : end - start;
+	let i, point, prev;
 
-		if (currentVM.skip) {
-			move = move || !spanGaps;
+	for (i = 0; i <= ilen; ++i) {
+		point = points[(start + (reverse ? ilen - i : i)) % count]._view;
+
+		if (point.skip) {
+			// If there is a skipped point inside a segment, spanGaps must be true
 			continue;
-		}
-		if (move) {
-			ctx.moveTo(currentVM.x, currentVM.y);
+		} else if (move) {
+			ctx.moveTo(point.x, point.y);
 			move = false;
-		} else if (vm.tension || steppedLine) {
-			lineMethod(ctx, previousVM, currentVM, false, steppedLine);
 		} else {
-			ctx.lineTo(currentVM.x, currentVM.y);
+			lineMethod(ctx, prev, point, reverse, vm.steppedLine);
 		}
-		previousVM = currentVM;
+
+		prev = point;
+	}
+
+	if (loop) {
+		point = points[(start + (reverse ? ilen : 0)) % count]._view;
+		lineMethod(ctx, prev, point, reverse, vm.steppedLine);
+	}
+
+	return !!loop;
+}
+
+function path(ctx, line, options) {
+	const segments = line.segments;
+	const ilen = segments.length;
+	let loop = line._loop;
+	for (let i = 0; i < ilen; ++i) {
+		loop &= pathSegment(ctx, line, segments[i], options);
 	}
+	return !!loop;
 }
 
 /**
  * Create path from points, grouping by truncated x-coordinate
  * Points need to be in order by x-coordinate for this to work efficiently
  * @param {CanvasRenderingContext2D} ctx - Context
  * @param {Point[]} points - Points defining the line
- * @param {boolean} spanGaps - Are gaps spanned over
+ * @param {Object} line - Line
  */
-function fastPath(ctx, points, spanGaps) {
-	let move = true;
+function fastPath(ctx, points, line, options) {
+	const ilen = points.length;
+	let {move = true, startIndex = 0, reverse = false} = options;
 	let count = 0;
 	let avgX = 0;
-	let index, vm, truncX, x, y, prevX, minY, maxY, lastY;
+	let i, index, point, truncX, prevX, minY, maxY, lastY;
 
-	for (index = 0; index < points.length; ++index) {
-		vm = points[index]._view;
+	for (i = 0; i < ilen; ++i) {
+		index = (startIndex + i) % ilen;
```

Comment:
Should we declare some of these variables where they're first used? E.g. `const index` here, etc.

Reply:
done